cli/internal/cmd/upgradeapply.go

```diff
@@ -186,37 +197,52 @@ func (u *upgradeApplyCmd) upgradeApply(cmd *cobra.Command, upgradeDir string, fl
 		return fmt.Errorf("applying migration for upgrading from v2.11: %w", err)
 	}
 
-	if err := u.confirmAndUpgradeAttestationConfig(cmd, conf.GetAttestationConfig(), idFile.MeasurementSalt, flags); err != nil {
+	if err := u.confirmAndUpgradeAttestationConfig(cmd, conf.GetAttestationConfig(), stateFile.ClusterValues.MeasurementSalt, flags); err != nil {
 		return fmt.Errorf("upgrading measurements: %w", err)
 	}
 
-	var infraState state.Infrastructure
+	// If infrastructure phase is skipped, we expect the new infrastructure
+	// to be in the Terraform configuration already. Otherwise, perform
+	// the Terraform migrations.
+	var postMigrationInfraState state.Infrastructure
 	if flags.skipPhases.contains(skipInfrastructurePhase) {
-		infraState, err = u.clusterShower.ShowInfrastructure(cmd.Context(), conf.GetProvider())
+		postMigrationInfraState, err = u.clusterShower.ShowInfrastructure(cmd.Context(), conf.GetProvider())
 		if err != nil {
-			return fmt.Errorf("getting infra state: %w", err)
+			return fmt.Errorf("getting Terraform state: %w", err)
 		}
 	} else {
-		infraState, err = u.migrateTerraform(cmd, conf, upgradeDir, flags)
+		postMigrationInfraState, err = u.migrateTerraform(cmd, conf, upgradeDir, flags)
 		if err != nil {
 			return fmt.Errorf("performing Terraform migrations: %w", err)
 		}
 	}
-	// reload idFile after terraform migration
-	// it might have been updated by the migration
-	if err := u.fileHandler.ReadJSON(constants.ClusterIDsFilename, &idFile); err != nil {
-		return fmt.Errorf("reading updated cluster ID file: %w", err)
+
+	// Merge the pre-upgrade state with the post-migration infrastructure values
+	if _, err := stateFile.Merge(
+		// temporary state with post-migration infrastructure values
+		state.New().SetInfrastructure(postMigrationInfraState),
+	); err != nil {
+		return fmt.Errorf("merging pre-upgrade state with post-migration infrastructure values: %w", err)
 	}
-	state := state.NewState(infraState)
-	// TODO(elchead): AB#3424 move this to updateClusterIDFile and correctly handle existing state when writing state
-	if err := u.fileHandler.WriteYAML(constants.StateFilename, state, file.OptOverwrite); err != nil {
+
+	// Write the post-migration state to disk
+	if err := stateFile.WriteToFile(u.fileHandler, constants.StateFilename); err != nil {
 		return fmt.Errorf("writing state file: %w", err)
 	}
+
+	cmd.Printf("Infrastructure migrations applied successfully and output written to: %s and %s\n"+
```

Comment:
Moving this outside of the `migrateTerraform` function will cause it to get printed every time `upgrade apply` is run. Even when the Terraform is skipped, or there are no updates to be done.

Reply:
True. I think we can spare the "migrations applied successfully" print in case of a manual infrastructure migration and leave the implementation of this simple. What do you think?

Reply:
You could just move this back into `migrateTerraform`

Reply:
This is the obvious way, but it would come at the consequence of no notice to the user that the state-file has been updated with the new infrastructure values in case of a manual Terraform migration (i.e. skipTerraformPhase)

Reply:
In case of manual Terraform migration, the state file changes would have been made by the user, so there shouldnt be a need to inform them about the changes